### PR TITLE
feat:[CDS-84546]: Harness Store Support for SAM Directory and Values Manifest

### PR DESCRIPTION
**Pull request by piyush-bhuwalka**
## Describe your changes

Harness Store Support for SAM Directory and Values Manifest

For more details, refer this doc:

https://harness.atlassian.net/wiki/spaces/DEV/pages/21551744043/Download+Manifest+From+Harness+Store+in+CD+Container+Step+Group

Cases tested:

- Aws Sam Directory Manifest in Harness Store and Values in Git
- Aws Sam Directory Manifest in Git and Values in Harness Store
- Both Aws Sam Directory Manifest and Values in Harness Store
- Both Aws Sam Directory Manifest and Values in Git
- Providing a folder as path for Aws Sam Directory Manifest

https://github.com/harness/harness-core/assets/100153994/070d81f4-45fd-4b08-a6bb-cbc89c776214

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56775)

### CODE REVIEW

**Comment by piyush-bhuwalka** [comment]
trigger uts

**Comment by piyush-bhuwalka** [comment]
trigger uts

**Comment by piyush-bhuwalka** [comment]
trigger ut6, ut7, ut8, ut9

**Comment by piyush-bhuwalka** [comment]
trigger uts

**Comment by piyush-bhuwalka** [comment]
trigger ut9

**Comment by piyush-bhuwalka** [comment]
trigger ut6, ut7, ut8, ut9

**Review by ivan-83 — APPROVED** [review]
:lgtm:

> Reviewed 8 of 11 files at r1, 2 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56775#-:-Nle5yag8p_ptTiHcKvH:bnfp4nl)* status: 11 file(s) to review. 1 of 2 approvals obtained.

**Comment by piyush-bhuwalka** [comment]
trigger uts

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/DownloadManifestsCommonHelper.java [hunk 1]
```diff
@@ -195,7 +197,7 @@ public StepElementParameters getDownloadS3StepElementParameters(
         .name(manifestOutcome.getIdentifier())
         .identifier(manifestOutcome.getIdentifier())
         .spec(downloadAwsS3StepParameters)
-        .timeout(ParameterField.createValueField("10m"))
+        .timeout(ParameterField.createValueField(DEFAULT_TIMEOUT))
```
Comment (PragyeshMishra01) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/DownloadManifestsCommonHelper.java:200]: shall we put timeout that user has configured for step?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/DownloadManifestsCommonHelper.java [hunk 2]
```diff
@@ -242,7 +244,7 @@ public StepElementParameters getDownloadHarnessStoreStepElementParameters(
         .name(manifestOutcome.getIdentifier())
         .identifier(manifestOutcome.getIdentifier())
         .spec(downloadHarnessStoreStepParameters)
-        .timeout(ParameterField.createValueField("10m"))
+        .timeout(ParameterField.createValueField(DEFAULT_TIMEOUT))
```
Comment (PragyeshMishra01) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/DownloadManifestsCommonHelper.java:247]: same

**Review by PragyeshMishra01 — APPROVED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/DownloadManifestsCommonHelper.java (hunk 2 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/DownloadManifestsCommonHelper.java:247]: Lets keep this as it is for now. Brought this in discussion with PM here

https://harness.atlassian.net/browse/CDS-86560

File: 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/DownloadManifestsCommonHelper.java (hunk 1 above)
Comment (piyush-bhuwalka) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/DownloadManifestsCommonHelper.java:200]: Lets keep this as it is for now. Brought this in discussion with PM here

https://harness.atlassian.net/browse/CDS-86560

**Review by ivan-83 — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56775)* status: 12 file(s) to review. 2. unresolved discussions. 

___
*[`125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/DownloadManifestsCommonHelper.java` line 247 at r4](https://reviewable.io/reviews/harness/harness-core/56775#gh-1427566059:-NlgedfX8zhOnRzA4CsB:bnjd9ws) ([raw file](https://github.com/harness/harness-core/blob/a75158270419b7d492af99badf36f750d15859cc/125-cd-nextgen/src/main/java/io/harness/cdng/aws/sam/DownloadManifestsCommonHelper.java#L247)):*
<details><summary><i>Previously, piyush-bhuwalka wrote…</i></summary><blockquote>

Lets keep this as it is for now. Brought this in discussion with PM here

https://harness.atlassian.net/browse/CDS-86560
</blockquote></details>

Good catch, pls address this @piyush-bhuwalka otherwise will come to us as CFD.

**Comment by piyush-bhuwalka** [comment]
trigger review

**Comment by acasian-harness** [comment]
hey @piyush-bhuwalka 

1.  Where do we validate that only 1 file is allowed. with the following store type?
2. Where are we validating that this actually contains data in order to avoid null pointer?
`valuesManifestOutcomeStore.getFiles().getValue().get(0)).getFileName().toString())`

**Comment by piyush-bhuwalka** [comment]
We validate it in 125-cd-nextgen/src/main/java/io/harness/cdng/plugininfoproviders/ServerlessV2PluginInfoProviderHelper.java line No 211 . This step runs before the actual step, so it would be caught.

I'll add it at the specified place as well for double check

**Pull request merged**